Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3735,6 +3737,8 @@ auto AirIRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signa
             });
         });
     }
+    // The call could have been to another WebAssembly instance, and / or could have modified our Memory.
+    restoreWebAssemblyGlobalState(RestoreCachedStackLimit::Yes, m_info.memory, currentInstance, m_currentBlock);
```

Comment:
Here, we restore the state unconditionally, since the callee might have tail-called an import. Is this too much of a cost? Otherwise we could check for that information statically.

Reply:
Note for future patch: we could check the transitive closure of your tail calls to see if this is needed

Reply:
This might be a perf regression

Reply:
I think this is very likely to be a perf regression. Justin has an idea on how to make this better.

restoring the wasm global state does 4 loads and 1 store.

Reply:
Justin's idea is this: you only restore the state if it calls something that transitively tail calls an import.

So if foo calls bar, bar tail calls baz, and baz is an import, then when foo calls bar, it'll restore state, even though bar is in the same module as foo.

But if baz happens to be in the same module as bar, it won't have this bit set, and we can omit restoring the state.

Another example, if foo calls bar, bar tail calls baz, baz tail calls jaz, and foo, bar, baz are in module 1, and jaz is in module 2, we still need to restore state at foo calling bar.